src/target/at32f43x.c

```diff
@@ -0,0 +1,376 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * This file implements AT32F43x target specific functions for detecting
+ * the device, providing the XML memory map and Flash memory programming.
+ *
+ * References:
+ * ARTERY doc - RM_AT32F435_437_EN_V2.04.pdf
+ *   Reference manual - AT32F435/437 Series Reference Manual
+ */
+
+#include "general.h"
+#include "target.h"
+#include "target_internal.h"
+#include "cortexm.h"
+
+static bool at32f43_flash_erase(target_flash_s *flash, target_addr_t addr, size_t len);
+static bool at32f43_flash_write(target_flash_s *flash, target_addr_t dest, const void *src, size_t len);
+static bool at32f43_mass_erase(target_s *target);
+
+/* Flash memory controller register map */
+#define AT32F435_FLASH_REG_BASE 0x40023c00U
+#define AT32F435_FLASH_UNLOCK   (AT32F435_FLASH_REG_BASE + 0x04U)
+#define AT32F435_FLASH_STS      (AT32F435_FLASH_REG_BASE + 0x0cU)
+#define AT32F435_FLASH_CTRL     (AT32F435_FLASH_REG_BASE + 0x10U)
+#define AT32F435_FLASH_ADDR     (AT32F435_FLASH_REG_BASE + 0x14U)
+//#define AT32F435_FLASH_UNLOCK2  (AT32F435_FLASH_REG_BASE + 0x44U)
+//#define AT32F435_FLASH_STS2     (AT32F435_FLASH_REG_BASE + 0x4cU)
+//#define AT32F435_FLASH_CTRL2    (AT32F435_FLASH_REG_BASE + 0x50U)
+//#define AT32F435_FLASH_ADDR2    (AT32F435_FLASH_REG_BASE + 0x54U)
+
+#define FLASH_BANK1_REG_OFFSET 0x00U
+#define FLASH_BANK2_REG_OFFSET 0x40U
+//#define FLASH_BANK_SPLIT_2K    0x08080000U
+//#define FLASH_BANK_SPLIT_4K    0x08200000U
+
+/* Flash registers bit fields */
+#define FLASH_CTRL_FPRGM   (1U << 0U)
+#define FLASH_CTRL_SECERS  (1U << 1U)
+#define FLASH_CTRL_BANKERS (1U << 2U)
+#define FLASH_CTRL_ERSTR   (1U << 6U)
+#define FLASH_CTRL_OPLK    (1U << 7U)
+//#define FLASH_CTRL_BLKERS  (1U << 3U)
+/* CTRL bits [8:11] are reserved, parallellism x8/x16/x32 (don't care) */
+
+/* OBF is BSY, ODF is EOP */
+#define FLASH_STS_OBF     (1U << 0U)
+#define FLASH_STS_PRGMERR (1U << 2U)
+#define FLASH_STS_ODF     (1U << 5U)
+
+#define FLASH_SR_BSY (1U << 16U)
+
+#define KEY1 0x45670123U
+#define KEY2 0xcdef89abU
+
+#define DBGMCU_IDCODE 0xe0042000U
+
+#define AT32F4x_IDCODE_SERIES_MASK 0xfffff000U
+#define AT32F4x_IDCODE_PART_MASK   0x00000fffU
+#define AT32F43_SERIES_4K          0x70084000U
+#define AT32F43_SERIES_2K          0x70083000U
+
+typedef struct at32f43_flash {
+	target_flash_s flash;
+	target_addr_t bank_split; /* Address of first page of bank 2 */
+} at32f43_flash_s;
+
+static void at32f43_add_flash(target_s *const target, const target_addr_t addr, const size_t length,
+	const size_t pagesize, const target_addr_t bank_split)
+{
+	if (length == 0)
+		return;
+
+	at32f43_flash_s *sf = calloc(1, sizeof(*sf));
+	if (!sf) { /* calloc failed: heap exhaustion */
+		DEBUG_ERROR("calloc: failed in %s\n", __func__);
+		return;
+	}
+
+	target_flash_s *flash = &sf->flash;
```

Comment:
We suggest renaming `flash` here to `target_flash` so you can rename `sf` to `flash` and avoid introducing yet more short variable names we need to fix.

Reply:
Can do. There are 9 other uses of `sf`, should I do something about them, too?
May I inquire why were they originally called `t`, `f` and `sf` in legacy driver code?

Reply:
We couldn't hope to know what was going on in Gareth's mind nor other contributor's, but as the short variable names have lead to readability issues and confusion and as it's a clang-tidy lint, we've been gradually fixing this over the last couple of releases

Reply:
Rename complete.